### DISCUSSION TITLE
Only format inside certain nodes

### DISCUSSION THREAD

**Discussion by uncenter** [topiary/topiary#791] [Q&A]
I'm trying to write a formatter for Topiary, or using Topiary, for a templating language - [Tera](https://keats.github.io/tera/docs/). I've written the following Tree-sitter grammar for the language: https://github.com/uncenter/tree-sitter-tera.

Tera templating logic is contained within "tags" and the external content (which is ignored by the language, arbitrary text) is marked under the `content` node. A content node usually has non-whitespace characters but could contain just spaces and newlines between tags, such as:

![CleanShot 2024-11-07 at 09 58 38](https://github.com/user-attachments/assets/cdc53a25-d6f1-459a-8f5d-e37876f9ef99)

I would like Topiary to completely ignore whitespace contained within `content` nodes, but it appears to be stripping it.

The example above:

```
{# A comment with some whitespace (newlines) after it. #}

{{ an_expression_tag_with_a_variable }}
```

is formatted to:

```
{# #}{{ an_expression_tag_with_a_variable }}
```

(Why the contents of the comments are being removed is another question entirely.)

Is this possible to solve? What do you suggest?

**Comment by uncenter** [comment]
Just noticed I have forgotten to provide my topiary queries:

```scheme
[
  "{{"
  "{%"
  "{#"
] @append_space

[
  "}}"
  "%}"
  "#}"
] @prepend_space

[
  "+"
  "-"
  "~"
  "%"
  "*"
  "/"

  "|"

  "=="
  "!="
  "<"
  ">"
  ">="
  "<="

  "if"
  "elif"
  "else"
  "endif"

  "for"
  "endfor"

  "include"
  "import"
  "as"
  "extends"

  "in"
  "and"
  "or"
  "not"
  "is"

  "break"
  "continue"

  "set"
  "set_global"
  "filter"
  "endfilter"
  "block"
  "endblock"
  "macro"
  "endmacro"
  "raw"
  "endraw"
] @prepend_space @append_space

(assignment_expression "=" @prepend_space @append_space)

; Remove trailing comma from last element of single line array.
(array
  ("," @delete . "]")(#single_line_only!)
)

("," @append_spaced_softline)
```

**Reply by Xophmeister** [reply]
Does adding content nodes as a leaf do the trick? Something like:

```scm
(content) @leaf
```

Topiary will, in general, consume all whitespace unless you tell it otherwise. However, "leaf" nodes are left untouched, so this _might_ be a workaround for you.

**Reply by uncenter** [reply]
I tried that earlier and just tried adding that snippet you gave to confirm and yeah no changes on my end, still stripping the whitespace.
